pkg/services/ngalert/accesscontrol/rules.go

```diff
@@ -34,106 +31,157 @@ func NewRuleService(ac accesscontrol.AccessControl) *RuleService {
 	}
 }
 
-// HasAccess returns true if the user has all permissions specified by the evaluator
-func (r *RuleService) HasAccess(ctx context.Context, user identity.Requester, evaluator accesscontrol.Evaluator) bool {
-	result, err := r.ac.Evaluate(ctx, user, evaluator)
-	if err != nil { // this is how accesscontrol.HasAccess works. //TODO change when AuthorizeDatasourceAccessForRule can return errors
-		logger.FromContext(ctx).Error("Failed to evaluate access control", "error", err)
-		return false
+// HasAccess returns true if the identity.Requester has all permissions specified by the evaluator. Returns error if access control backend could not evaluate permissions
+func (r *RuleService) HasAccess(ctx context.Context, user identity.Requester, evaluator accesscontrol.Evaluator) (bool, error) {
+	return r.ac.Evaluate(ctx, user, evaluator)
+}
+
+// HasAccessOrError returns nil if the identity.Requester has enough permissions to pass the accesscontrol.Evaluator. Otherwise, returns authorization error that contains action that was performed
+func (r *RuleService) HasAccessOrError(ctx context.Context, user identity.Requester, evaluator accesscontrol.Evaluator, action func() string) error {
+	has, err := r.HasAccess(ctx, user, evaluator)
+	if err != nil {
+		return err
 	}
-	return result
+	if !has {
+		return NewAuthorizationErrorWithPermissions(action(), evaluator)
+	}
+	return nil
 }
 
-// AuthorizeDatasourceAccessForRule checks that user has access to all data sources declared by the rule
-func (r *RuleService) AuthorizeDatasourceAccessForRule(ctx context.Context, user identity.Requester, rule *models.AlertRule) bool {
-	for _, query := range rule.Data {
-		if query.QueryType == expr.DatasourceType || query.DatasourceUID == expr.DatasourceUID || query.
-			DatasourceUID == expr.OldDatasourceUID {
-			continue
-		}
-		if !r.HasAccess(ctx, user, accesscontrol.EvalPermission(datasources.ActionQuery, datasources.ScopeProvider.GetResourceScopeUID(query.DatasourceUID))) {
-			return false
+// getRulesReadEvaluator constructs accesscontrol.Evaluator that checks all permission required to read all provided rules
+func (r *RuleService) getRulesReadEvaluator(rules ...*models.AlertRule) accesscontrol.Evaluator {
+	return r.getRulesQueryEvaluator(rules...)
+}
+
+// getRulesQueryEvaluator constructs accesscontrol.Evaluator that checks all permissions to query data sources used by the provided rules
+func (r *RuleService) getRulesQueryEvaluator(rules ...*models.AlertRule) accesscontrol.Evaluator {
+	added := make(map[string]struct{}, 2)
+	evals := make([]accesscontrol.Evaluator, 0, 2)
+	for _, rule := range rules {
+		for _, query := range rule.Data {
+			if query.QueryType == expr.DatasourceType || query.DatasourceUID == expr.DatasourceUID || query.
```

Comment:
I know this wasn't technically changed in this PR, but could consider using `query.IsExpression()` here. Could even do a very small refactor to remove the unused `error` return so it can be inlined.

Reply:
I would rather get rid of the `query.IsExpression` because it makes a dependency on `expr` package, and keep the logic that binds `expr` and `ng.models` packages in 3rd package.

This check is not necessary, though, and can be replaced by similar logic as https://github.com/grafana/grafana/blob/580477bf8ee0eebd5fc74d9dc274d795aa510461/pkg/services/ngalert/eval/eval.go#L298-L307

to facilitate checks for `MLNode` in the future.